File: frontend/src/types/quest.ts
export interface QuestSummary {
  id: string
  title: string
  phase: number
  order: number
  xp_reward: number
  completed: boolean
  unlocked: boolean
}

export interface TestCase {
  type: string
  expected: string
  input_data: string | null
  description: string
}

export interface QuestTask {
  id: string
  description: string
  starter_code: string
  test_cases: TestCase[]
  hints: string[]
  order: number
}

export interface QuestData {
  id: string
  title: string
  phase: number
  order: number
  story: string
  content: string
  tasks: QuestTask[]
  challenge: QuestTask | null
  xp_reward: number
  coin_reward: number
}

export interface TestResult {
  passed: boolean
  description: string
  expected: string
  actual: string
  error: string | null
}

export interface TaskSubmitResult {
  task_id: string
  passed: boolean
  test_results: TestResult[]
  execution_time: number
  stdout: string
  stderr: string
}

File: backend/schemas/quest.py
from pydantic import BaseModel


class TestCase(BaseModel):
    type: str = "output_match"
    expected: str
    input_data: str | None = None
    description: str = ""


class QuestTask(BaseModel):
    id: str
    description: str
    starter_code: str
    test_cases: list[TestCase]
    hints: list[str] = []
    order: int = 1


class QuestData(BaseModel):
    id: str
    title: str
    phase: int
    order: int
    story: str
    content: str
    tasks: list[QuestTask]
    challenge: QuestTask | None = None
    xp_reward: int = 100
    coin_reward: int = 50


class QuestSummary(BaseModel):
    id: str
    title: str
    phase: int
    order: int
    xp_reward: int
    completed: bool = False
    unlocked: bool = False


class TaskSubmit(BaseModel):
    task_id: str
    code: str


class TestResult(BaseModel):
    passed: bool
    description: str
    expected: str
    actual: str
    error: str | None = None


class TaskSubmitResult(BaseModel):
    task_id: str
    passed: bool
    test_results: list[TestResult]
    execution_time: float
    stdout: str = ""
    stderr: str = ""


class QuestProgress(BaseModel):
    quest_id: str
    completed: bool
    completed_tasks: list[str]
